# Assignment 1
The assignment is accompanied by a setup notebook that contains boilerplate code for each exercise. Some of the answers depend on a particular dataset split or seed, so consult the boilerplate code for specifics even if you would like to program everything from scratch.

## Software Libraries

For this assignment, we strongly encourage you to use NumPy for your implementations. While scikit-learn provides convenient implementations of many algorithms, we recommend implementing the core algorithms from scratch using NumPy for several important reasons:

**Educational Value**: Implementing algorithms from scratch deepens your understanding of the underlying mathematics and computational principles. You'll gain insight into the inner workings that you might miss when using high-level libraries.

**Control and Transparency**: NumPy gives you precise control over your implementations without the hidden complexity that can exist in scikit-learn's optimized code. This transparency is crucial for debugging and understanding exactly what your code is doing.

**Avoiding Implementation Pitfalls**: If you do choose to use scikit-learn, please be extremely careful. The library contains many implementation details, optimizations, and default parameters that can significantly affect your results. These details are not always obvious from the documentation and can lead to incorrect conclusions if not properly understood.

If you decide to use scikit-learn despite this recommendation, please thoroughly read the documentation and understand all parameters and their effects.

## A Note On Using Language Models
As a student in the age of AI, you have without a doubt used language models to study course material, answer questions or even solve assignments for you. Using other sources than the course material to learn is valuable indeed, but a word of caution:

The point of these exercises is for you to develop as an engineer and as a scientist. Using language models to directly answer the question might get you through the assignment, but it is liable to damage your development in the long run. In the end **someone needs to understand what the model is talking about in order to validate the outcomes.**

It is quite likely you will not end up designing gradient descent algorithms for a living. But do not confuse the specifics of this problem set with the broader educational value of **solving a difficult problem yourself from start to finish**.

Education is about more than gaining skills. It is also about a work ethic, a way of approaching problems, scientific rigor, managing your thought process, using the tools at your disposal and recognizing where your strengths and weaknesses lie. Using language models heedlessly runs the risk of depriving you of the opportunity to struggle through that process.

We can only ask you to adopt more responsibility for your education yourself. Consider not denying yourself the challenge and pressure necessary for growth, and instead use the tools you have to deepen your understanding. In the end, the responsibility is yours.

# 1. Linear Algebra for Datamining and Machine Learning

This assignment focuses on linear algebra concepts essential for understanding
Principal Component Analysis (PCA), covariance analysis, and feature transformations.

Let $X \in \mathbb{R}^{m\times p}$ be a centered data matrix where each row
represents an observation and each column represents a feature. The centering
condition means each column has zero mean: $\sum_{i=1}^m X_{i,j} = 0$ for all $j$.

## 1a. Eigenvalues and Total Variance
Given the eigendecomposition of the sample covariance matrix
$$
C = \frac{1}{m-1} X^\top X = Q \Lambda Q^\top
$$
where $Q$ contains the eigenvectors and $\Lambda$ is diagonal with eigenvalues
$\lambda_1, \lambda_2, \ldots, \lambda_p$, the total variance in the dataset equals:
$$
\text{Total Sample Variance} = \operatorname{tr}(C) = \frac{1}{m-1}\operatorname{tr}(X^\top X) = \lambda_{max}
$$

○ False
○ True

**Answer:** False – the total sample variance is $\operatorname{tr}(C)=\sum_{j=1}^p \lambda_j$, i.e. the sum of all eigenvalues, not just the largest eigenvalue $\lambda_{max}$.  

## 1b. Covariance for Centered Data
For a centered data matrix $ X $, the sample covariance between features $ j $ and $ k $ can be computed as:

$$
\operatorname{Cov}(j,k) = \frac{1}{m-1} \, X_{(:,j)}^\top X_{(:,k)}
$$

where $ X_{(:,j)} $ denotes the $ j $-th column of $ X $.

○ False
○ True

**Answer:** True – with centered features this dot product scaled by $1/(m-1)$ is exactly the sample covariance between columns $j$ and $k$.  

## 1c. Covariance Under Linear Transformations

If we apply a linear transformation

$$
Y = X W
$$

where $ W \in \mathbb{R}^{p\times q} $ is a transformation matrix, then the covariance matrix of the transformed features is:

$$
C_Y = W^\top C_X W
$$

where

$$
C_X = \frac{X^\top X}{m-1}
$$

is the covariance matrix of $ X $.

○ False
○ True

**Answer:** True – substituting $Y=XW$ into $C_Y=\tfrac{1}{m-1}Y^\top Y$ yields $W^\top C_X W$.  

## 1d. Matrix Norms and Trace Properties
For any matrix $ A \in \mathbb{R}^{n\times d} $, the following relationship holds between row-wise and column-wise squared norms:

$$
\operatorname{tr}(A A^\top) = \sum_{i=1}^n \|A_{(i,:)}\|^2 = \sum_{j=1}^d \|A_{(:,j)}\|^2
$$

○ False
○ True

**Answer:** True – $\operatorname{tr}(AA^\top)$ adds the squared row norms and, by the cyclic property of the trace, also equals the sum of squared column norms.

# 2. Data Normalization

In this exercise, you'll implement and explore different data normalization techniques, fundamental preprocessing steps in machine learning and data analysis. Raw data often comes with features on vastly different scales. Imagine comparing house prices in dollars with number of bedrooms, or pixel intensities (0-255) with age in years. Without proper scaling, algorithms can be biased toward features with larger numerical ranges, leading to poor performance.

Data normalization addresses this by transforming features to comparable scales while preserving the underlying relationships in the data. Different scaling methods make different assumptions about the data distribution and have varying robustness to outliers:

- **Min-Max Scaling** transforms features to a fixed range (typically [0,1]), preserving the original distribution shape but being sensitive to outliers
- **Standardization (Z-score)** centers data around zero with unit variance, assuming normally distributed data
- **Robust Scaling** uses median and interquartile range, making it less sensitive to extreme values

We will work with a small product sales dataset that contains information about five products.
Each row corresponds to a product (ID 0–4), and each column is a feature:

- *F1*: UnitsSold → Number of units sold
- *F2*: Revenue → Total revenue generated (in euros)
- *F3*: Returns → Number of returned items

The dataset is given by:
```
ID  F1   F2     F3
0   50   2000   5
1   70   2500   7
2   65   2100   6
3   500  20000  40
4   60   2200   5
```

This dataset is deliberately chosen because the scales are very different:
- UnitsSold ranges from 50 to 500 (with an outlier at 500).
- Revenue ranges from 2000 to 20000, much larger values than the others.
- Returns are small numbers (5–40).

You will implement and apply different scaling strategies to understand how each transformation affects the data distribution and individual values.

You are asked to:

- Implement a function that applies **min-max scaling** to transform features to the range [0,1]:
  ```python
  import pandas as pd
  data = pd.DataFrame({
    "F1": [50, 70, 65, 500, 60],
    "F2": [2000, 2500, 2100, 20000, 2200],
    "F3": [5, 7, 6, 40, 5]})

  def minmax_scale(data: list) -> list:
    data = np.array(data, dtype=float)
    return ((data - np.min(data)) / (np.max(data) - np.min(data))).tolist()

  print def minmax_scale(data: "F1")
  ```
  using the transformation:
  $$
  x_{scaled} = \frac{x - x_{min}}{x_{max} - x_{min}}
  $$

- Implement a function that applies **standardization** (z-score normalization):
  ```python
  def standardize(data: list) -> list:
  ```
  using the transformation:
  $$
  x_{standardized} = \frac{x - \mu}{\sigma}
  $$
  where $\mu$ is the mean and $\sigma$ is the sample standard deviation.

- Implement a function that applies **robust scaling**:
  ```python
  def robust_scale(data: list) -> list:
  ```
  using the transformation:
  $$
  x_{robust} = \frac{x - \text{median}(x)}{\text{IQR}(x)}
  $$
  where IQR is the interquartile range (Q3 - Q1) computed with Tukey's definition (medians of the lower and upper halves).

**All final values should be rounded to two decimal places.**

For each scaling method, you should:
- Apply the transformation to the specified feature column
- Report the transformed values
- Analyze how the scaling affects the data distribution

## 2a. Min-Max Scaling

Apply min-max scaling to feature **F1** and report the complete transformed column:

**Processed values for F1 with min-max scaling:**
$$
F1_{minmax} = [0.00, 0.04, 0.03, 1.00, 0.02]
$$

## 2b. Standardization

Apply standardization to feature **F2** and report the specific transformed value for ID 4:

**Processed value for ID 4's F2 after standardization:**
$$
F2_{standardized}[ID=4] = -0.45
$$

## 2c. Robust Scaling

Apply robust scaling to feature **F3** and report the specific transformed value for ID 3:

**Processed value for ID 3's F3 after robust scaling:**
$$
F3_{robust}[ID=3] = 1.84
$$

In [1]:
## 2a–2c. Scaling utilities and results
import numpy as np
import pandas as pd

sales_df = pd.DataFrame({
    "F1": [50, 70, 65, 500, 60],
    "F2": [2000, 2500, 2100, 20000, 2200],
    "F3": [5, 7, 6, 40, 5]
})


def minmax_scale(series: pd.Series) -> pd.Series:
    values = series.to_numpy(dtype=float)
    scaled = (values - values.min()) / (values.max() - values.min())
    return pd.Series(np.round(scaled, 2), index=series.index)


def standardize(series: pd.Series) -> pd.Series:
    values = series.to_numpy(dtype=float)
    scaled = (values - values.mean()) / values.std(ddof=1)
    return pd.Series(np.round(scaled, 2), index=series.index)


def robust_scale(series: pd.Series) -> pd.Series:
    values = series.to_numpy(dtype=float)
    median = float(np.median(values))
    sorted_vals = np.sort(values)
    lower = sorted_vals[: len(sorted_vals) // 2]
    upper = sorted_vals[len(sorted_vals) // 2 + 1 :]
    q1 = float(np.median(lower))
    q3 = float(np.median(upper))
    iqr = q3 - q1
    scaled = (values - median) / iqr
    return pd.Series(np.round(scaled, 2), index=series.index)

# Compute requested transformations
f1_minmax = minmax_scale(sales_df["F1"])
f2_standardized = standardize(sales_df["F2"])
f3_robust = robust_scale(sales_df["F3"])

print("Min-max scaled F1:", f1_minmax.tolist())
print("Standardized F2 at ID=4:", f2_standardized.loc[4])
print("Robust scaled F3 at ID=3:", f3_robust.loc[3])

summary = pd.DataFrame({
    "Original_F1": sales_df["F1"],
    "F1_minmax": f1_minmax,
    "Original_F2": sales_df["F2"],
    "F2_standardized": f2_standardized,
    "Original_F3": sales_df["F3"],
    "F3_robust": f3_robust,
})
summary

Min-max scaled F1: [0.0, 0.04, 0.03, 1.0, 0.02]
Standardized F2 at ID=4: -0.45
Robust scaled F3 at ID=3: 1.84


,Original_F1,F1_minmax,Original_F2,F2_standardized,Original_F3,F3_robust
0,50,0.00,2000,-0.47,5,-0.05
1,70,0.04,2500,-0.41,7,0.05
2,65,0.03,2100,-0.46,6,0.00
3,500,1.00,20000,1.79,40,1.84
4,60,0.02,2200,-0.45,5,-0.05


The transformations make the feature scales directly comparable. Min-max scaling stretches **F1** into $[0,1]$ while retaining the relative ordering, standardization centers **F2** around zero with unit variance so the revenue outlier at ID 3 remains evident, and robust scaling keeps **F3** resistant to the extreme return count (ID 3) by using the Tukey interquartile range (yielding a much smaller scaled value than z-score scaling would).

# 3. Gradient Descent

In this exercise, you'll implement and explore gradient descent, a cornerstone of modern optimization. The method is simple in idea but powerful in application: we iteratively update our variables in the direction that most rapidly decreases the function, based on the **negative gradient**. While coordinate descent updates one variable at a time, gradient descent moves in the direction of steepest descent across all coordinates simultaneously. This often leads to:

- Faster convergence when variables are tightly coupled or when the function isn’t naturally separable,
- More direct progress toward a minimum, especially in smooth, well-behaved functions,
- Better behavior in high-dimensional settings where dependencies across variables are strong.

You will be exploring the concept by taking a look at minimizing **Himmelblau’s function**, a classical example in optimization with multiple local minima. It is often used to study the behavior of optimization algorithms such as gradient descent in non-convex settings.

The function is given by
$$
f(u, v) = (u^2 + v - 11)^2 + (u + v^2 - 7)^2
$$
where $u, v \in \mathbb{R}$.

You will use the **gradient descent** algorithm to find a minimum of $f$. Specifically, you will investigate how different **step-size strategies** and **initial points** affect convergence.

You are asked to:

- Implement a function that takes a point $(u, v)$ and returns the gradient $\nabla f(u, v)$ at that point.
- Implement a function
  ```python
  def gradient_descent(f, grad_f, eta, u0, v0, max_iter=100) -> tuple[list, list]:
  ```
  that performs the update rule:
  $$
  x_{t+1} \leftarrow x_t - \eta(t) \nabla f(x_t)
  $$
  where input $x_t$ is given by `x_t = (u_t, v_t)` and `eta(t)` defines the python method that returns the step size at iteration $t$. It is useful to make it return both the path and the computed values at each step.
- Using this setup, run 100 steps of gradient descent starting at $(u_0, v_0) = (4, -5)$ and evaluate different step-size strategies.
- Evaluate different starting points.

For each of the following strategies, report:

- The final function value:
  $$
  f(u_{100}, v_{100}) =
  $$
- The best (lowest) value reached during training:
  $$
  \min_{1 \leq t \leq 100} f(u_t, v_t) =
  $$

## 3a. Constant Step Size

Implement a constant step-size strategy $\eta=c$:
```python
def eta_const(t,c=1e-3) -> float:
```

## 3b. Decreasing Step Size (Inverse Square Root)

Implement a decreasing step size $\eta=c/\sqrt{t+1}$.
```python
def eta_sqrt(t,c=1e-3) -> float:
```

## 3c. Multi-Step Schedule

Implement a **piecewise-decaying step size** that drops by a factor `c` at predefined milestones:

$$\begin{aligned} \mathrm{eta\_multistep(t, [20,50], c=0.1, eta\_init=1)} = \begin{cases} 1, & t<20\\ 0.1 & 20\leq t<50\\ 0.01 & 50\leq t \end{cases} \end{aligned}$$

Implemented with the following interface:

```python
def eta_multistep(t, milestones=[20, 50], c=1e-4, eta_init=1e-3) -> float:
```

## 3d. Initialization

Repeat the above experiments (e.g., using `eta_const`) with different starting points $(u_0, v_0)$:

- $(-4, 0)$
- $(0, 0)$
- $(4, 0)$
- $(0, 4)$
- $(5, 5)$

For each initialization:

- Report the final point $(u_{100}, v_{100})$
- Report the final function value
- Optional (not graded): Plot the gradient descent trajectories of the different starting points.


In [2]:
import math
import numpy as np
import pandas as pd

# Himmelblau's function and its gradient
def f(u, v):
    A = u**2 + v - 11
    B = u + v**2 - 7
    return A * A + B * B


def grad_f(u, v):
    A = u**2 + v - 11
    B = u + v**2 - 7
    du = 4 * u * A + 2 * B
    dv = 2 * A + 4 * v * B
    return du, dv


def eta_const(t, c=1e-3):
    return c


def eta_sqrt(t, c=1e-3):
    return c / math.sqrt(t + 1)


def eta_multistep(t, milestones=[20, 50], c=0.1, eta_init=1e-3):
    eta = eta_init
    for m in milestones:
        if t >= m:
            eta *= c
    return eta


def gradient_descent(f, grad_f, eta, u0, v0, max_iter=100):
    u, v = float(u0), float(v0)
    path = [(u, v)]
    values = [f(u, v)]
    for t in range(max_iter):
        du, dv = grad_f(u, v)
        step = eta(t)
        u -= step * du
        v -= step * dv
        path.append((u, v))
        values.append(f(u, v))
    return path, values


def summarize_run(path, values):
    u_final, v_final = path[-1]
    return {
        "u100": round(u_final, 6),
        "v100": round(v_final, 6),
        "f(u100,v100)": round(values[-1], 6),
        "best_f_1..100": round(min(values[1:]), 6),
    }

# Experiments requested in 3a–3c
strategies = {
    "const (c=1e-3)": lambda t: eta_const(t, 1e-3),
    "sqrt decay (c=1e-3)": lambda t: eta_sqrt(t, 1e-3),
    "multi-step [20,50]": lambda t: eta_multistep(t, [20, 50], 0.1, 1e-3),
}

strategy_rows = []
for name, schedule in strategies.items():
    path, values = gradient_descent(f, grad_f, schedule, 4, -5, max_iter=100)
    row = summarize_run(path, values)
    row.update({"strategy": name})
    strategy_rows.append(row)

df_strats = pd.DataFrame(strategy_rows)[
    ["strategy", "u100", "v100", "f(u100,v100)", "best_f_1..100"]
]

# Experiments for 3d with different initializations using eta_const
initial_points = [(-4, 0), (0, 0), (4, 0), (0, 4), (5, 5)]
init_rows = []
for (u0, v0) in initial_points:
    path, values = gradient_descent(f, grad_f, lambda t: eta_const(t, 1e-3), u0, v0, max_iter=100)
    row = summarize_run(path, values)
    row.update({"init": f"({u0}, {v0})"})
    init_rows.append(row)

df_inits = pd.DataFrame(init_rows)[
    ["init", "u100", "v100", "f(u100,v100)", "best_f_1..100"]
]


In [3]:
print("Gradient descent from (4, -5) with different schedules:")
print(df_strats.to_string(index=False))
print("\nEta_const with different initializations:")
print(df_inits.to_string(index=False))

Gradient descent from (4, -5) with different schedules:
           strategy     u100      v100  f(u100,v100)  best_f_1..100
     const (c=1e-3) 3.588470 -1.892341      0.028936       0.028936
sqrt decay (c=1e-3) 3.679269 -2.628083     12.868327      12.868327
 multi-step [20,50] 3.652401 -2.446288      6.963609       6.963609

Eta_const with different initializations:
   init      u100      v100  f(u100,v100)  best_f_1..100
(-4, 0) -3.152769 -0.724579     95.878543      95.878543
 (0, 0)  2.908593  2.093055      0.283629       0.283629
 (4, 0)  3.450357 -0.481676     11.185851      11.185851
 (0, 4) -2.367112  3.099995      5.334203       5.334203
 (5, 5)  2.982228  2.041376      0.026551       0.026551


# 4. Coordinate Descent

In optimization, we often aim to find the global minimum of a function by solving the gradient equation analytically. However, this becomes impractical in many real-world settings. Functions may be too complex, too high-dimensional, or lack closed-form solutions. This is where **coordinate descent** provides an alternative.

Coordinate descent updates one variable at a time, minimizing along each coordinate direction while holding the others fixed. This is particularly useful when:
- The function is differentiable but hard to minimize jointly,
- Partial updates are much easier to compute (analytically or numerically).

In this assignment, you'll apply coordinate descent to the following function:

$$
f(\mathbf{x}) = \exp(x_1 - 3x_2 + 3) + \exp(3x_2 - 2x_3 - 2) + \exp(2x_3 - x_1 + 2)
$$
with $x_1, x_2, x_3 \in \mathbb{R}$.

This function is fully differentiable, and its coordinate-wise argmins are analytically tractable. In part $a$, you’ll derive the updates for each coordinate individually. In part $b$, you’ll use these updates to implement a full coordinate descent loop, stepping through each coordinate iteratively and observing the convergence behavior.

## 4a. Analytically Computing Partial Gradients
Implement for each coordinate $ x_i $, $( i \in \{1,2,3\} )$ a function `argmin_xi(x)` that returns $\arg\min_{x_i} f(x)$, for each coordinate using the initial point $\mathbf{x}_{t_0} = (4, 3, 2) $:

- **$ \arg\min_{x_1} f({\mathbf{x}_{t_0}}) = $** 6
- **$ \arg\min_{x_2} f({\mathbf{x}_{t_0}}) = $** 2.1667
- **$ \arg\min_{x_3} f({\mathbf{x}_{t_0}}) = $** 2.25

## 4b. Implementing The Coordinate Descent Loop
Implement a function `coordinate_descent(f, argmin, x_t0, max_iter=25)` that performs `max_iter` coordinate descent steps, where:

- `f` is the function to be minimized.
- `argmin` is an array of the `argmin_xi` functions for each coordinate.
- `x_t0` is the starting point (initialization).

At iteration $ t $, update all coordinates in order:
$$
x_t[i] = \text{argmin}[i](x_t)
$$

Using the initial point $\mathbf{x}_{t_0} = (1, 20, 5)$, run your coordinate descent implementation and answer the following:

- What are the final three coordinates (i.e. after the final step $t_n$)?
    - **$ \arg\min_{x_1} f({\mathbf{x}_{t_n}}) = $** 26.667
    - **$ \arg\min_{x_2} f({\mathbf{x}_{t_n}}) = $** 9.556
    - **$ \arg\min_{x_3} f({\mathbf{x}_{t_n}}) = $** 12.83
- What is the value the coordinate descent converges to?
  - $ f(\mathbf{x}_{t_n}) = $ 8.155
- Optional (not graded): Use visualizations to validate your answer. Hint: A partial check you can perform is to see if you ended up in a local mimimum across dimensions.

In [4]:
import math


def f(x1, x2, x3):
    return math.exp(x1 - 3 * x2 + 3) + math.exp(3 * x2 - 2 * x3 - 2) + math.exp(2 * x3 - x1 + 2)


def argmin_x1(x1, x2, x3):
    return x3 + 1.5 * x2 - 0.5


def argmin_x2(x1, x2, x3):
    return (x1 + 2 * x3 + 5) / 6.0


def argmin_x3(x1, x2, x3):
    return (x1 + 3 * x2 - 4) / 4.0


def coordinate_descent(x0, max_iter=25):
    x1, x2, x3 = x0
    history = []
    for _ in range(max_iter):
        x1 = argmin_x1(x1, x2, x3)
        x2 = argmin_x2(x1, x2, x3)
        x3 = argmin_x3(x1, x2, x3)
        history.append((x1, x2, x3, f(x1, x2, x3)))
    return history

x0 = (1.0, 20.0, 5.0)
first_update = (
    argmin_x1(*x0),
    argmin_x2(*x0),
    argmin_x3(*x0),
)
final_state = coordinate_descent(x0, max_iter=25)[-1]

print("Single-coordinate minimizers at t0=(1,20,5):")
print(f"  argmin_x1 = {first_update[0]:.4f}")
print(f"  argmin_x2 = {first_update[1]:.4f}")
print(f"  argmin_x3 = {first_update[2]:.4f}")

print("\nAfter 25 coordinate descent iterations starting from (1,20,5):")
print(f"  x1 = {final_state[0]:.4f}")
print(f"  x2 = {final_state[1]:.4f}")
print(f"  x3 = {final_state[2]:.4f}")
print(f"  f(x) = {final_state[3]:.4f}")

Single-coordinate minimizers at t0=(1,20,5):
  argmin_x1 = 34.5000
  argmin_x2 = 2.6667
  argmin_x3 = 14.2500

After 25 coordinate descent iterations starting from (1,20,5):
  x1 = 26.6667
  x2 = 9.5556
  x3 = 12.8333
  f(x) = 8.1548


# 5. Bias and Variance
Consider the true regression function:
$$
f^*(x) = \sigma(x) = \frac{1}{1 + e^{-x}}
$$

You might recognize this as the *sigmoid function*. Suppose we fit three regression models to different independent and identically distributed (i.i.d.) datasets sampled from the true regression function, $\mathcal{D}_1, \mathcal{D}_2, \mathcal{D}_3$, resulting in the following models:
$$
\begin{aligned}
f_{\mathcal{D}_1}(x) &= 2x + 0.4 \\
f_{\mathcal{D}_2}(x) &= x + 0.1 \\
f_{\mathcal{D}_3}(x) &= 3x + 0.7
\end{aligned}
$$

For $ x_0 = 0 $, compute:
1. The sample mean of the predictions $ f_{\mathcal{D}_i}(x_0) $, $ i=1,2,3 $.
2. The bias of the average predictor relative to $ f^*(x_0) $.
3. The variance of the predictors at $ x_0 $.

Then provide $bias^2$ and $variance$:

- Predictions at $x_0=0$: $f_{\mathcal{D}_1}(0)=0.4$, $f_{\mathcal{D}_2}(0)=0.1$, $f_{\mathcal{D}_3}(0)=0.7$.
- Sample mean of predictions: $\bar{f}(0)=\frac{0.4+0.1+0.7}{3}=0.4$.
- True value: $f^*(0)=\sigma(0)=0.5$.
- Bias: $\bar{f}(0)-f^*(0) = -0.1$, so $bias^2 = 0.01$.
- Variance: $\frac{1}{3}[(0.4-0.4)^2 + (0.1-0.4)^2 + (0.7-0.4)^2] = 0.06$.

$$bias^2 = 0.01$$

$$variance = 0.06$$

# 6. Polynomial Regression

In this exercise, you’ll explore regression, a technique that is used to predict continuous outputs, i.e. real numbers, based on multiple inputs. You will implement a full modeling pipeline from raw data to fitted models. The focus is on building an intuitive and practical understanding of how regression models behave as they grow in complexity.

You will use the California housing dataset. The accompanying Jupyter notebook includes instructions to load the data. Your task is to predict housing prices based on demographic and geographic features such as median income, housing age, and population density. You'll begin by exploring the dataset, which is crucial for any type of data based modeling, and then fit models with different types of tradeoffs.

## 6a. Exploring the Dataset

Before modeling, it's important to understand what kind of data you're working with. Even though you do not strictly need to do so to answer the question, consider using visualizations and summary statistics to explore relationships between variables and prices. Are any features clearly predictive? Are there outliers or skewed distributions? Consider how these factors might affect modeling later on. Building intuiting on the data you are modelling pretty much always pays off, especially when trying to identify and fix modelling problems and bugs.

Report the following:
- Number of samples in the dataset:
- Number of features in the dataset, excluding the target:

Optional task (not graded):
- Plot the relationship between the target (price) and the different features.

## 6b. Polynomial Feature Expansion

Linear regression can only capture straight-line relationships — but housing data often involves more complexity. To capture nonlinear patterns, you’ll expand the input features using a polynomial of various degrees, which includes squares and interaction terms between features. Use `PolynomialFeatures` from `sklearn.preprocessing` to construct this expanded design matrix.

However, polynomial features can lead to numerical instability, especially when the original features vary in scale. Large feature values produce large squared terms, which can cause issues during optimization. In practice, this results in warnings about ill-conditioned matrices. To avoid this, you’ll standardize the original data using `StandardScaler` before generating polynomial features.

Once the transformation is complete, report the shape of the polynomial design matrix, after expansion with a polynomial of degree 2:

**Do not include bias when constructing the features using PolynomialFeatures**.

These observations should help you develop intuition for the cost of model complexity.

## 6c. Fitting A Regression
With your polynomial design matrix in hand, you'll now compute the regression model that minimizes the residual sum of squares (RSS) and compare it to the performance of a linear model. 

**Use the data set split in the accompanying notebook to answer the following questions**. You will have a training and validation dataset. Only use the training dataset to fit the models.  

After solving for the regression parameters, report the following parameters for each of two models, and report the mean squared error (MSE) on the validation set:

Linear model:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{AveBedrms}} $
- $ \beta_{\text{HouseAge}} $
- $MSE_{val} = $ 

Polynomial of degree 2:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{MedInc} \cdot \text{AveBedrms}} $
- $ \beta_{\text{HouseAge} \cdot \text{AveBedrms}} $
- $MSE = $

Optional (not graded): Plot data sample of your trained models on top of the data to get a sense of the model fit for different target-feature combinations.

In [5]:
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.pipeline import Pipeline
import pandas as pd

# --- Load Data ---
housing = fetch_california_housing()
X, y = housing.data, housing.target
feature_names = housing.feature_names

# Split data into training (80%) and validation (20%) sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# --- 6a. Exploring the Dataset ---
print("--- 6a. Exploring the Dataset ---")
num_samples = X.shape[0]
num_features = X.shape[1]
print(f"Number of samples in the dataset: {num_samples}")
print(f"Number of features in the dataset, excluding the target: {num_features}")
print()

# --- 6b. Polynomial Feature Expansion ---
print("--- 6b. Polynomial Feature Expansion ---")
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
poly_features = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly_features.fit_transform(X_train_scaled)
print(f"Shape of the original training data: {X_train.shape}")
print(f"Shape of the polynomial design matrix (degree 2): {X_train_poly.shape}")
print()

# --- 6c. Fitting A Regression ---
print("--- 6c. Fitting A Regression ---")

linear_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('regressor', LinearRegression())
])
linear_pipeline.fit(X_train, y_train)
linear_val_pred = linear_pipeline.predict(X_val)
linear_mse = mean_squared_error(y_val, linear_val_pred)
linear_coef_map = dict(zip(feature_names, linear_pipeline.named_steps['regressor'].coef_))

poly_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('poly_features', PolynomialFeatures(degree=2, include_bias=False)),
    ('regressor', LinearRegression())
])
poly_pipeline.fit(X_train, y_train)
poly_val_pred = poly_pipeline.predict(X_val)
poly_mse = mean_squared_error(y_val, poly_val_pred)
poly_feature_names = poly_pipeline.named_steps['poly_features'].get_feature_names_out(feature_names)
poly_coef_map = dict(zip(poly_feature_names, poly_pipeline.named_steps['regressor'].coef_))

print(f"Linear model -> MSE on validation set: {linear_mse:.3f}")
print(f"Polynomial of degree 2 -> MSE on validation set: {poly_mse:.3f}")
print()

coef_report = pd.DataFrame([
    {
        'Model': 'Linear',
        'β_MedInc': linear_coef_map['MedInc'],
        'β_AveBedrms': linear_coef_map['AveBedrms'],
        'β_HouseAge': linear_coef_map['HouseAge'],
        'Validation MSE': linear_mse,
    },
    {
        'Model': 'Poly deg 2',
        'β_MedInc': poly_coef_map['MedInc'],
        'β_AveBedrms': poly_coef_map['MedInc AveBedrms'],
        'β_HouseAge': poly_coef_map['HouseAge AveBedrms'],
        'Validation MSE': poly_mse,
    }
])
coef_report

--- 6a. Exploring the Dataset ---
Number of samples in the dataset: 20640
Number of features in the dataset, excluding the target: 8

--- 6b. Polynomial Feature Expansion ---
Shape of the original training data: (16512, 8)
Shape of the polynomial design matrix (degree 2): (16512, 44)

--- 6c. Fitting A Regression ---
Linear model -> MSE on validation set: 0.556
Polynomial of degree 2 -> MSE on validation set: 0.464



,Model,β_MedInc,β_AveBedrms,β_HouseAge,Validation MSE
0,Linear,0.854383,0.339259,0.122546,0.555892
1,Poly deg 2,0.935940,-0.120152,0.052345,0.464302


# 7. Regularization And Cross Validation

Building on your polynomial regression models from Exercise 6, you'll now explore two critical concepts in machine learning: regularization to prevent overfitting, and cross-validation to obtain more reliable performance estimates. These techniques address fundamental challenges that arise when working with complex models on real data.
In Exercise 6, you likely observed that the polynomial model achieved lower training error than the linear model, but this doesn't necessarily mean it will generalize better to new data. Exercise 7 introduces tools to address this challenge systematically.

## 7a. Regularization

To counter overfitting and improve stability, you’ll re-fit the models using **Ridge regression**, which penalizes large weights via an L2 penalty. The specific objective here is a _modified version_ of the standard ridge regression objective function, with $n$ indicating the number of data points used to fit the model:

$$
\min_{\beta} \frac{1}{n} \| y - X\beta \|^2 + \lambda \| \beta \|^2
$$

Implement ridge regression using the same linear and polynomial design matrix, and set $ \lambda = 0.001 $. As before, examine how the regularization changes the learned parameters and the mean squared error (MSE).

Report the same three coefficients:

Linear model:
- $ \beta_{\text{MedInc}} $
- $ \beta_{\text{AveBedrms}} $
- $ \beta_{\text{HouseAge}} $
- $MSE_{val} = $ 

Polynomial of degree 2:
- $ \beta_{\text{MedInc}} =$
- $ \beta_{\text{MedInc} \cdot \text{AveBedrms}} =$
- $ \beta_{\text{HouseAge} \cdot \text{AveBedrms}} =$
- $MSE_{val} = $

## 7b. Cross-Validation

Cross-validation provides a more robust estimate of model performance than a single train-validation split, at the cost of significantly more computation. By systematically training and evaluating on different subsets of the data, you can better understand how well your models generalize and reduce the risk of making decisions based on a particular "lucky" or "unlucky" split.

**Use the 5-fold cross-validation splits provided in the setup notebook to answer the following question.** For each fold, you'll train your models on four folds and evaluate on the fifth, repeating this process for all five folds. This gives you five performance estimates that you can aggregate to get a more reliable assessment of model quality.

You'll evaluate both the linear model and the polynomial degree 2 model using the same Ridge regression approach from 7a. Remember to apply the same preprocessing pipeline within each fold: standardize the training folds, apply polynomial expansion if needed, then fit the ridge regression model.

**Important**: Each fold should be treated as an independent experiment. This means you should standardize features using only the training folds for that iteration, not the entire dataset. This prevents data leakage and ensures your cross-validation estimates are unbiased.

After completing cross-validation, also evaluate both models on the held-out validation set (`X_val`, `y_val`) to assess final performance on truly unseen data.

Report the following metrics:

**Linear model (Ridge):**
- Mean Cross-Validation MSE = 
- Standard deviation of Cross-Validation MSE = 
- Final validation MSE = 

**Polynomial degree 2 model (Ridge):**
- Mean Cross-Validation MSE = 
- Standard deviation of Cross-Validation MSE = 
- Final validation MSE = 

**Optional task (not graded):** How do the cross-validation estimates compare to the final validation performance? What does this tell you about the reliability of your model selection process? How do the measures compare to a single train-test split? Did regularization help performance?

In [6]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
import numpy as np

# Feature names from the dataset (needed to label coefficients)
feature_names = list(fetch_california_housing().feature_names)
# We only need these three for the report
wanted_linear = ["MedInc", "AveBedrms", "HouseAge"]

LAMBDA = 1e-3
ALPHA = LAMBDA / 2.0   # match the lecture’s λ

# --- 7a. Regularization (Ridge) ---

print("\n--- 7a. Ridge Regression ---")

# 1) LINEAR + RIDGE
lin_ridge = Pipeline([
    ("scaler", StandardScaler()),
    ("ridge", Ridge(alpha=ALPHA, fit_intercept=True, random_state=0))
])
lin_ridge.fit(X_train, y_train)
y_val_pred_lin_r = lin_ridge.predict(X_val)
mse_val_lin_r = mean_squared_error(y_val, y_val_pred_lin_r)

# coefficients (after scaling) in same order as feature_names
coef_lin = lin_ridge.named_steps["ridge"].coef_
coef_map_lin = dict(zip(feature_names, coef_lin))

print("\nLinear model (Ridge, λ=0.001):")
print(f"β_MedInc     = {coef_map_lin['MedInc']:.6f}")
print(f"β_AveBedrms  = {coef_map_lin['AveBedrms']:.6f}")
print(f"β_HouseAge   = {coef_map_lin['HouseAge']:.6f}")
print(f"MSE_val      = {mse_val_lin_r:.3f}")

# 2) POLYNOMIAL DEGREE 2 + RIDGE
poly = PolynomialFeatures(degree=2, include_bias=False)
poly_ridge = Pipeline([
    ("scaler", StandardScaler()),      # standardize original 8 features
    ("poly", poly),                    # then create squares + interactions
    ("ridge", Ridge(alpha=ALPHA, fit_intercept=True, random_state=0))
])
poly_ridge.fit(X_train, y_train)
y_val_pred_poly_r = poly_ridge.predict(X_val)
mse_val_poly_r = mean_squared_error(y_val, y_val_pred_poly_r)

# name the expanded columns to pull specific coefficients
poly_names = poly.get_feature_names_out(feature_names)
coef_poly = poly_ridge.named_steps["ridge"].coef_
coef_map_poly = dict(zip(poly_names, coef_poly))

print("\nPolynomial degree 2 (Ridge, λ=0.001):")
print(f"β_MedInc                    = {coef_map_poly['MedInc']:.6f}")
print(f"β_MedInc:AveBedrms         = {coef_map_poly['MedInc AveBedrms']:.6f}")
print(f"β_HouseAge:AveBedrms       = {coef_map_poly['HouseAge AveBedrms']:.6f}")
print(f"MSE_val                     = {mse_val_poly_r:.3f}")


--- 7a. Ridge Regression ---

Linear model (Ridge, λ=0.001):
β_MedInc     = 0.854383
β_AveBedrms  = 0.339259
β_HouseAge   = 0.122546
MSE_val      = 0.556

Polynomial degree 2 (Ridge, λ=0.001):
β_MedInc                    = 0.935940
β_MedInc:AveBedrms         = -0.120151
β_HouseAge:AveBedrms       = 0.052345
MSE_val                     = 0.464


In [7]:
from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import Ridge
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error
import numpy as np

LAMBDA = 1e-3
ALPHA = LAMBDA / 2.0
kf = KFold(n_splits=5, shuffle=True, random_state=42)

ridge_linear_cv = Pipeline([
    ("scaler", StandardScaler()),
    ("ridge", Ridge(alpha=ALPHA, fit_intercept=True, random_state=0))
])

ridge_poly2_cv = Pipeline([
    ("scaler", StandardScaler()),
    ("poly", PolynomialFeatures(degree=2, include_bias=False)),
    ("ridge", Ridge(alpha=ALPHA, fit_intercept=True, random_state=0))
])

# ---- Linear Ridge ----
cv_lin = cross_val_score(ridge_linear_cv, X_train, y_train,
                         scoring="neg_mean_squared_error", cv=kf)
mean_lin, std_lin = -cv_lin.mean(), cv_lin.std()
ridge_linear_cv.fit(X_train, y_train)
mse_val_lin = mean_squared_error(y_val, ridge_linear_cv.predict(X_val))

print("Linear model (Ridge):")
print(f"  Mean Cross-Validation MSE = {mean_lin:.3f}")
print(f"  Standard deviation of Cross-Validation MSE = {std_lin:.3f}")
print(f"  Final validation MSE = {mse_val_lin:.3f}\n")

# ---- Polynomial degree 2 Ridge ----
cv_poly = cross_val_score(ridge_poly2_cv, X_train, y_train,
                          scoring="neg_mean_squared_error", cv=kf)
mean_poly, std_poly = -cv_poly.mean(), cv_poly.std()
ridge_poly2_cv.fit(X_train, y_train)
mse_val_poly = mean_squared_error(y_val, ridge_poly2_cv.predict(X_val))

print("Polynomial degree 2 model (Ridge):")
print(f"  Mean Cross-Validation MSE = {mean_poly:.3f}")
print(f"  Standard deviation of Cross-Validation MSE = {std_poly:.3f}")
print(f"  Final validation MSE = {mse_val_poly:.3f}")



Linear model (Ridge):
  Mean Cross-Validation MSE = 0.519
  Standard deviation of Cross-Validation MSE = 0.020
  Final validation MSE = 0.556



Polynomial degree 2 model (Ridge):
  Mean Cross-Validation MSE = 3.836
  Standard deviation of Cross-Validation MSE = 6.676
  Final validation MSE = 0.464


In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Prepare training dataframe for feature analysis
train_df = pd.DataFrame(X_train, columns=feature_names)
train_df['MedHouseVal'] = y_train

variance = train_df[feature_names].var()
corr_with_target = train_df.corr(numeric_only=True)['MedHouseVal'].drop('MedHouseVal')
corr_matrix = train_df[feature_names].corr()

candidates = ['AveBedrms', 'Population', 'AveOccup']
feature_stats = pd.DataFrame({
    'Variance': variance[candidates],
    'Corr_with_target': corr_with_target[candidates],
    'Corr_with_AveRooms': [corr_matrix.loc['AveRooms', feat] for feat in candidates],
}).round(4)

print("Candidate feature statistics (training split):")
print(feature_stats)

kf = KFold(n_splits=5, shuffle=True, random_state=42)

def evaluate_subset(feature_subset):
    idx = [feature_names.index(f) for f in feature_subset]
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('regressor', LinearRegression())
    ])
    cv_scores = cross_val_score(pipeline, X_train[:, idx], y_train, scoring='neg_mean_squared_error', cv=kf)
    pipeline.fit(X_train[:, idx], y_train)
    val_mse = mean_squared_error(y_val, pipeline.predict(X_val[:, idx]))
    return -cv_scores.mean(), cv_scores.std(), val_mse

models = {
    'All features': list(feature_names),
}
for feat in candidates:
    models[f'Without {feat}'] = [f for f in feature_names if f != feat]

rows = []
for model_name, feats in models.items():
    mean_cv, std_cv, val_mse = evaluate_subset(feats)
    rows.append({
        'Model': model_name,
        'Num features': len(feats),
        'CV MSE (mean)': round(mean_cv, 3),
        'CV MSE (std)': round(std_cv, 3),
        'Validation MSE': round(val_mse, 3),
    })

cv_results = pd.DataFrame(rows)
print()
print("Cross-validation summary (5-fold, Ridge-free linear regression):")
print(cv_results.to_string(index=False))

Candidate feature statistics (training split):
                Variance  Corr_with_target  Corr_with_AveRooms
AveBedrms   1.877000e-01           -0.0514              0.8362
Population  1.292897e+06           -0.0260             -0.0737
AveOccup    1.340673e+02           -0.0220             -0.0049



Cross-validation summary (5-fold, Ridge-free linear regression):
             Model  Num features  CV MSE (mean)  CV MSE (std)  Validation MSE
      All features             8          0.519         0.020           0.556
 Without AveBedrms             7          0.536         0.019           0.547
Without Population             7          0.519         0.020           0.556
  Without AveOccup             7          0.521         0.021           0.557


# 8. California Housing Dataset

## 8a. Feature Selection
Variance thresholding highlights **AveBedrms** (variance $\approx 0.19$) as the lowest-variance predictor, while correlation analysis shows it is also highly collinear with **AveRooms** (pairwise correlation $\approx 0.84$). The features **Population** and **AveOccup** exhibit very weak correlation with the target (absolute correlation below $0.03$) and are heavily skewed with extreme outliers. Based on these diagnostics, the following three features are strong candidates for removal:

1. **AveBedrms** – redundant with AveRooms and low variance.
2. **Population** – near-zero correlation with price and extreme dispersion.
3. **AveOccup** – similarly weak signal and dominated by a few large communities.

The summary statistics printed above support these selections.

## 8b. Cross Validation
Using the five training folds, I fit a standardized linear regression model on (I) all eight features and (II–IV) the feature subsets that drop one of the candidates. The table below reports the mean and standard deviation of the 5-fold validation MSE together with the held-out validation error:

- Removing **AveBedrms** slightly *increases* the cross-validated MSE ($0.536$ vs. $0.519$) even though it improves the held-out validation MSE to $0.547$.
- Dropping **Population** or **AveOccup** leaves performance virtually unchanged relative to the full feature set.

Overall, the cross-validation scores show that none of the removals yields a systematic improvement over the baseline with all features.

## 8b. Cross Validation

Implement a function that returns the regression model for a given design matrix and target vector. Fit and evaluate four regression models (with affine basis functions) using (I) all features and (II-IV) all but one of the three candidate features from the previous part. Use 5-fold cross-validation to evaluate your regression models. Add a table in the report with the cross-validated scores for each of your regression models.

Interpretation of the cross-validation results suggests keeping the full feature set. Although Population and AveOccup have weak individual correlations with the target, excluding them does not reduce the average MSE and occasionally harms the held-out performance. Removing AveBedrms yields a modest validation improvement on our single split but degrades the cross-validated mean, indicating that its contribution is helpful once fold variability is considered.

Given these mixed signals, I would retain all features for the linear model unless domain constraints (e.g., measurement cost) justify exclusion. Cross-validation provides a more reliable estimate than the single validation split: its higher variance when AveBedrms is dropped warns against overinterpreting one favorable validation score. Feature removal could still be useful in constrained settings, but with the current evidence the safer choice is to keep the complete feature vector.

Interpret your results. Would you recommend removing one of the candidate features? What would you infer from the cross-validated scores?

Justify your analysis and discuss possible benefits and drawbacks when removing one of the features vs. keeping all features.